### PR TITLE
Avoid GPU0 CUDA context growth from checkpoint common-step broadcast

### PR DESCRIPTION
**Pull request by yhgalaxy**
# What does this PR do ?

Avoid persistent GPU0 CUDA context growth during distributed checkpoint save by keeping the shared optimizer `common_step` value on CPU before it is inserted into checkpoint common state.

`common_step` is a scalar shared by optimizer parameter states. In some optimizer paths it may be a CUDA scalar tensor. Distributed checkpoint code can treat common checkpoint state as Python object state and broadcast it with `torch.distributed.broadcast_object_list` for validation/coordination. If rank 0's common object contains a CUDA tensor, peer ranks can deserialize that object onto the serialized device, commonly `cuda:0`, which may initialize persistent CUDA contexts on GPU0 for processes whose real local training device is not GPU0.

This change avoids putting a CUDA tensor into the common object state. Tensor steps are detached, cloned, and moved to CPU; non-tensor steps keep the previous deep-copy behavior. The runtime optimizer state remains unchanged, and the checkpoint schema still contains the same `common_step` field.

This is a small bug fix with a narrow impact: it reduces checkpoint-induced GPU0 memory side effects without changing training behavior.

## Issue tracking

For PRs from open-source community contributors:

- **New features**: a linked issue is **required**. Please open a [feature request](https://github.com/NVIDIA/Megatron-LM/issues/new?template=feature_request.md) and reference it here before submitting the PR.
- **Small updates (bug fixes, minor improvements)**: a linked issue is **recommended** and will accelerate the PR review process.

Linked issue: N/A

### Pre-checks

- [x] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [x] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

Validation performed:

- `python3 -m py_compile megatron/core/optimizer/optimizer.py tests/unit_tests/test_optimizer.py`
- `python3 -m pytest tests/unit_tests/test_optimizer.py -k "checkpoint_common_step or sharded_state_dict_keeps_common_step" -q`

### Code review

Feel free to message or comment the [@mcore-oncall](https://github.com/orgs/NVIDIA/teams/mcore-oncall) to help accelerate your merge into main. The less complex your PR is, the faster it will be approved and merged!

All PRs start as **draft**. If you open a non-draft PR, it will be automatically converted to draft.

#### Step 1: Mark PR as "Ready for Review"

1. When your PR is ready, click **Ready for Review**.
2. An oncall reviewer is auto-assigned and expert reviewers are notified based on your changes.
   - Some PRs may jump straight to step 2. This is determined by `.github/CODEOWNERS`.

:warning: Only mark as ready once merge-conflicts are resolved and the CI is passing.
Final Review might get declined if these requirements are not fulfilled.

#### Step 2: Final Review

For PRs that change `megatron/core`, once all expert reviewers have approved, the `Final Review` label is applied **automatically** and final reviewers are assigned.

For PRs outside `megatron/core`, this step is skipped.

#### Step 3: Approved

Once all required reviewers have approved, the `Approved` label is applied **automatically**.

### Merge

Any member of [mcore-engineers](https://github.com/orgs/NVIDIA/teams/mcore-engineers) will be able to merge your PR.

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### CODE REVIEW

File: megatron/core/optimizer/optimizer.py [hunk 1]
```diff
@@ -369,6 +369,13 @@ def _restore_common_per_param_step(state_dict: Dict, step: Union[int, torch.Tens
         for param_idx, param_state in state_dict['state'].items():
             param_state['step'] = copy.deepcopy(step)
 
+    @staticmethod
+    def _checkpoint_common_step(step: Union[int, torch.Tensor]) -> Union[int, torch.Tensor]:
+        """Keep shared optimizer step safe for checkpoint common-state broadcasts."""
+        if torch.is_tensor(step):
+            return step.detach().cpu().clone()
```
[the remark this answers is not in the document]
Comment (yhgalaxy) [reply megatron/core/optimizer/optimizer.py:376]: Added regression coverage in `tests/unit_tests/test_optimizer.py` for the shared helper and both sharded optimizer checkpoint paths. The tests assert that tensor `step` values are emitted as CPU `common_step` tensors for FP32 and FP16 optimizer sharded state dicts.

**Comment by Connor-XY** [comment]
Hello @yhgalaxy, thanks for tracking down the GPU0 CUDA context growth around checkpoint common-step broadcast.

I checked the diff and the fix is targeted at the optimizer checkpoint metadata path: the common step is converted through a CPU scalar before the broadcast, so checkpoint save/load does not accidentally create or grow a CUDA context on GPU0 just to move this scalar state.

I’m routing this to @kunlunl for the distributed-optimizer/checkpoint ownership path.

**Comment by guihong-nv** [comment]
@yhgalaxy Please resolve the failing tests.
